spark/src/test/java/org/apache/iceberg/actions/TestExpireSnapshotsAction.java

```diff
@@ -0,0 +1,780 @@
+    return spark.createDataFrame(records, ThreeColumnRecord.class).coalesce(1);
+  }
+
+  private void writeDF(Dataset<Row> df, String mode) {
+    df.select("c1", "c2", "c3")
+        .write()
+        .format("iceberg")
+        .mode(mode)
+        .save(tableLocation);
+  }
+
+  private void checkExpirationResults(Long expectedDatafiles, Long expectedManifestsDeleted,
+      Long expectedManifestListsDeleted, ExpireSnapshotsActionResult results) {
+
+    Assert.assertEquals("Incorrect number of manifest files deleted",
+        expectedManifestsDeleted, results.getManifestFilesDeleted());
+    Assert.assertEquals("Incorrect number of datafiles deleted",
+        expectedDatafiles, results.getDataFilesDeleted());
+    Assert.assertEquals("Incorrect number of manifest lists deleted",
+        expectedManifestListsDeleted, results.getManifestListsDeleted());
+  }
+
+  @Test
+  public void testFilesCleaned() throws Exception {
+    Dataset<Row> df = buildDF(RECORDS);
+
+    writeDF(df, "append");
+
+    List<Path> expiredDataFiles = Files
+        .list(tableDir.toPath().resolve("data").resolve("c1=1"))
+        .collect(Collectors.toList());
+
+    Assert.assertEquals("There should be a data file to delete but there was none.",
+        2, expiredDataFiles.size());
+
+    writeDF(df, "overwrite");
+    writeDF(df, "append");
+
+    long end = System.currentTimeMillis();
+    while (end <= table.currentSnapshot().timestampMillis()) {
+      end = System.currentTimeMillis();
+    }
+
+    ExpireSnapshotsActionResult results =
+        Actions.forTable(table).expireSnapshots().expireOlderThan(end).execute();
+
+    table.refresh();
+
+    Assert.assertEquals("Table does not have 1 snapshot after expiration", 1, Iterables.size(table.snapshots()));
+
+    for (Path p : expiredDataFiles) {
+      Assert.assertFalse(String.format("File %s still exists but should have been deleted", p),
+          Files.exists(p));
+    }
+
+    checkExpirationResults(1L, 2L, 2L, results);
+  }
+
+  @Test
+  public void dataFilesCleanupWithParallelTasks() throws IOException {
+
+    table.newFastAppend()
+        .appendFile(FILE_A)
+        .commit();
+
+    table.newFastAppend()
+        .appendFile(FILE_B)
+        .commit();
+
+    table.newRewrite()
+        .rewriteFiles(ImmutableSet.of(FILE_B), ImmutableSet.of(FILE_D))
+        .commit();
+    long thirdSnapshotId = table.currentSnapshot().snapshotId();
+
+    table.newRewrite()
+        .rewriteFiles(ImmutableSet.of(FILE_A), ImmutableSet.of(FILE_C))
+        .commit();
+    long fourthSnapshotId = table.currentSnapshot().snapshotId();
+
+    long t4 = System.currentTimeMillis();
+    while (t4 <= table.currentSnapshot().timestampMillis()) {
+      t4 = System.currentTimeMillis();
+    }
+
+    Set<String> deletedFiles = Sets.newHashSet();
+    Set<String> deleteThreads = ConcurrentHashMap.newKeySet();
+    AtomicInteger deleteThreadsIndex = new AtomicInteger(0);
+
```

Comment:
Manual metadata rewrite code removed from here since it wasn't really part of the test and I couldn't easily move those functions into Spark module